stratisproject/StratisBitcoinFullNode
# [SC] Experimental! An approach to nesting mappings and lists

Hey @rowandh @fassadlr,

Would be great if you could take a look at this. I've left all of the existing mapping and list implementations there, and just added a couple of methods and classes to demonstrate a way of nesting. If this is looking appealing for us I'll clean this up and make it into a formal PR next week. 

The idea:

When getting an object from a mapping or list, if it is a primitive or struct, then we check in persistent state at `${this.name}[{key}]`. If the object we are getting is a mapping or list (i.e. nested), then we return a new mapping/list, which has the name= `${this.name}[{key}]`. Obviously what this means is when we retrieve from this new mapping / list, the persistentstate gets all actually happen from `{originalName}[{key1}][{key2}]` (e.g. in persistentstate it might go to `mates[jordan][0]` for a `IScMapping<IScList<string>>("mates")` !! (See _ScMapping_ and _ScList_)

As for serialization, if we come up against a mapping or list in serialization, we just store the name to the db, and will load the mapping or list back when deserializing. (See PersistentStateSerializer).

How to review:
* Check out ScMapping and ScList, especially the part with the generic identification and construction.
* Check out PersistentStateSerializer
* Check out the contract: Mappings.cs, and then the tests that run on it, in SmartContractExecutorTests
* Excuse the couple of hacks I've put around the place.
